test/functional/p2p_headers_sync_with_minchainwork.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test that we reject low difficulty headers to prevent our block tree from filling up with useless bloat"""
+
+from test_framework.test_framework import BitcoinTestFramework
+
+import time
+
+NODE1_BLOCKS_REQUIRED = 15
+NODE2_BLOCKS_REQUIRED = 2047
+
+
+class RejectLowDifficultyHeadersTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+        # Node0 has no required chainwork; node1 requires 15 blocks on top of the genesis block; node2 requires 2047
+        self.extra_args = [["-minimumchainwork=0x0"], ["-minimumchainwork=0x1f"], ["-minimumchainwork=0x1000"]]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def test_chains_sync_when_long_enough(self):
+        self.log.info("Generate blocks on the node with no required chainwork")
+        self.generate(self.nodes[0], NODE1_BLOCKS_REQUIRED-2, sync_fun=self.no_op)
+
+        self.log.info("Verify nodes 1 and 2 have no new headers in their headers tree")
+        time.sleep(2)
+        for node in self.nodes[1:]:
```

Comment:
I agree with removing the `sleep()` since it's not a very reliable way to wait for things to happen in the functional tests (which are often run in parallel). Not sure about relying on `assert_debug_log`.

I'd recommend `wait_until` the headers are received, for example:

```suggestion
        self.log.info("Generate blocks on the node with no required chainwork")
        node1_recv_headers = self.nodes[1].getpeerinfo()[0]["bytesrecv_per_msg"]["headers"]
        node2_recv_headers = self.nodes[2].getpeerinfo()[0]["bytesrecv_per_msg"]["headers"]
        self.generate(self.nodes[0], NODE1_BLOCKS_REQUIRED-2, sync_fun=self.no_op)

        self.log.info("Verify nodes 1 and 2 have no new headers in their headers tree")
        self.wait_until(lambda: self.nodes[1].getpeerinfo()[0]["bytesrecv_per_msg"]["headers"] > node1_recv_headers)
        self.wait_until(lambda: self.nodes[2].getpeerinfo()[0]["bytesrecv_per_msg"]["headers"] > node2_recv_headers)
        for node in self.nodes[1:]:
```
(this is using `getpeerinfo()[0]` assuming node1 and node2 have node0 as their first peer)

Reply:
Ah, that's clever, but I think the problem is that the number of headers messages that come in after a call to `generate` is variable, based on the particular blocks that get INV'ed and the responses they generate.  

I took an approach of changing the logging to include the height of the chain, and then used the `assert_debug_log` approach to check for it.  This gets rid of all the sleeps and I think ought to work?